File: PlantreApp/app/src/main/java/com/example/plantreapp/myPlants/Plant.java
package com.example.plantreapp.myPlants;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;

import java.util.Objects;
import java.util.UUID;

public class Plant {
    private String id;
    private String name;
    private String description;

    public Plant(String name, String description) {
        this.id = UUID.randomUUID().toString();
        this.name = name;
        this.description = description;
    }

    public String getId() {
        return id;
    }

    /*public void setId(String id) {
        this.id = id;
    }*/

    public String getName() {
        return name;
    }

    /*public void setName(String name) {
        this.name = name;
    }*/

    public String getDescription() {
        return description;
    }

    /*public void setDescription(String description) {
        this.description = description;
    }*/

    @NonNull
    @Override
    public String toString() {
        return "Plant{" +
                "id='" + id + '\'' +
                ", name='" + name + '\'' +
                ", description='" + description + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Plant plant = (Plant) o;
        return id.equals(plant.id) &&
                name.equals(plant.name) &&
                description.equals(plant.description);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, name, description);
    }

    public static DiffUtil.ItemCallback<Plant> itemCallback = new DiffUtil.ItemCallback<Plant>() {
        @Override
        public boolean areItemsTheSame(@NonNull Plant oldItem, @NonNull Plant newItem) {
            return oldItem.getId().equals(newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull Plant oldItem, @NonNull Plant newItem) {
            return oldItem.equals(newItem);
        }
    };
}

File: PlantreApp/app/src/main/java/com/example/plantreapp/entities/Plant.kt
package com.example.plantreapp.entities

import androidx.recyclerview.widget.DiffUtil
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import java.util.*

@Entity
data class Plant (
    @PrimaryKey(autoGenerate = true) val uid:Int?,
    @ColumnInfo(name = "name") val name: String?,
    @ColumnInfo(name = "scientific_name") val scientificName: String?,
    @ColumnInfo(name = "picture") val picture: String?,
    @ColumnInfo(name = "description") val description: String?,
    @ColumnInfo(name = "stage") val stage: String?, // seed - seedling - mature
    @ColumnInfo(name = "seed_water_rate") val seedWaterRate: Int, // How often to water - in days 1-7, generally seedlings need a higher rate
    @ColumnInfo(name = "seedling_water_rate") val seedlingWaterRate: Int, // How often to water - in days 1-7, generally seedlings need a higher rate
    @ColumnInfo(name = "mature_water_rate") val matureWaterRate: Int, // How often to water - in days 1-7
    @ColumnInfo(name = "min_seed_moisture") val minSeedMoisture: Float, // minimum moisture level - seed
    @ColumnInfo(name = "max_seed_moisture")  val maxSeedMoisture: Float, // maximum moisture level - seed
    @ColumnInfo(name = "min_seedling_moisture") val minSeedlingMoisture: Float, // minimum moisture level - seedling
    @ColumnInfo(name = "max_seedling_moisture")  val maxSeedlingMoisture: Float, // maximum moisture level - seedling
    @ColumnInfo(name = "min_mature_moisture") val minMatureMoisture: Float, // minimum moisture level - Mature
    @ColumnInfo(name = "max_mature_moisture")  val maxMatureMoisture: Float // maximum moisture level - Mature
) {

    override fun hashCode(): Int {
        return Objects.hash(uid, name, description)
    }
    companion object {
        var itemCallback: DiffUtil.ItemCallback<Plant> = object : DiffUtil.ItemCallback<Plant>() {
            override fun areItemsTheSame(oldItem: Plant, newItem: Plant): Boolean {
                return oldItem.uid == newItem.uid
            }

            override fun areContentsTheSame(oldItem: Plant, newItem: Plant): Boolean {
                return oldItem == newItem
            }
        }
    }

}
